### ISSUE TITLE
Railway track blocks (junction restrictions)

### ISSUE THREAD

**Issue by dolanbaker** [CitiesSkylinesMods/TMPE#1001]
### Describe your idea
I suggest creating a sub-module for railway tracks that allows sections of track to be "blocked" as in classic railway signalling where only one train can ever be on a section of line at a time.
Trains are typically two or three track segments long and will stop only one segment behind another train if there is a queue, this can often mean that the train can stop while straddling a junction. This in turn means that any trains coming the other way that need to cross at the junction cannot as they are blocked by the train.  
If a train leaving a station or freight terminal is blocked in be a train straddling the junction, the next train cannot move into the station until this one has moved but it is blocked from moving, eventually causing the entire rail network to seize up!

The suggestion is to have a UI that allows the user to create an "absolute block" using a similar tool to the lane connector, where the start and of a block can be drawn on the railway track and the train will be stopped from entering that section until the previous train has completely exited the other end.

Such a tool would be useful at junctions to ensure that a train never enters the junction if it cannot completely clear the junction when it moves through.  As a further enhancement it can be linked to railway signals.

**Comment by krzychu124** [comment]
I thought about something like that and even more tools for train traffic management.
BTW, have you tried this mod: https://steamcommunity.com/sharedfiles/filedetails/?id=949504539?

**Comment by dolanbaker** [comment]
Cool, I will try it.

Thanks

**Comment by dolanbaker** [comment]
OK I tried the single tracks mod, but it still blocks at junctions, I have done a lot of work to minimise the locations where blocking can occur, but I still think that it would be a great enhancement if it was possible to mark sections of line for one train only at a time.

Maybe a tweak to the do not enter blocked junction feature.

**Comment by dolanbaker** [comment]
Looks like I have found a workaround, stretch the nodes as only one train can enter a node at any one time so by having streatched nodes either side of a junction prevents the trains blocking each other
![Long node](https://user-images.githubusercontent.com/68794550/92314071-cffb3300-efca-11ea-91fe-0579f7f78159.png)

**Comment by me22** [comment]
Cool workaround, @dolanbaker!  I'll have to try that.  (I assume you mean "stretch the segments", though?)

**Comment by dolanbaker** [comment]
Yeah, move the node to make the segment at least one train's length, then it will not allow another train to cross the junction unless it can fit into the segment without overhanging the junction(points).

**Comment by dolanbaker** [comment]
I found that the "node dissolver" in this mod **Elektrix's Road Tools 2.93** achieves the same result, but it would be better if there was a mod that could group several segments together and mark them as one for routing purposes.
Maybe something that uses some of the code from the bi-directional tracks mod could do the trick.

**Comment by xlf1024** [comment]
Another approach to this could be making trains check several segments ahead until the combined length is sufficient. This would adapt better to different train lengths and require no setup by the player thus being more beginner-friendly and eliminating the need for additional UI components. However, it would probably be less realistic given that most rail traffic irl is still block based.

Anyways, I'd love to see either (or both) implemented!

**Comment by dolanbaker** [comment]
> Another approach to this could be making trains check several segments ahead until the combined length is sufficient. This would adapt better to different train lengths and require no setup by the player thus being more beginner-friendly and eliminating the need for additional UI components. However, it would probably be less realistic given that most rail traffic irl is still block based.
> 
> Anyways, I'd love to see either (or both) implemented!
Yes, I agree 100% active block would be a great solution for basic rail management, as you say absolute block would require manual setting up.  Alternatively, a block can be between junctions (up to, say 20 segments), this would also work without user intervention.

**Comment by aubergine10** [comment]
The [Network Multitool](https://steamcommunity.com/sharedfiles/filedetails/?id=2560782729) mod also has a way to delete nodes which helps create big segments.

Also, I assume you've added vanilla stop signs on those train tracks (via the traffic junctions info view IIRC)?
